Adding generic gamelist filter funcionality for ES

pjft opened the pull request:
Specifically filters for genre, players, ratings and publisher/developer.

Made it fairly generic in order to accommodate any metadata rearchitecture, or Zigurana's kid-friendly/favorites changes.

Rebased with current ES from RetroPie. Hope it suits!

jrassa commented:
@pjft This is a great new feature, but I think it might need a little more work.

My first thought after reviewing the code is that it would be better if the filtering did not occur in BasicGameListView.  If you look at [how sorting works](https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/guis/GuiGamelistOptions.cpp#L60), the options menu triggers a re-sort on the root FileData object and then tells the gamelist that a change has occurred.  The gamelist then repopulates the list.  It is MVC like and keeps things decoupled.

With your current implementation, if we move the metadata to sqlite, BasicGameListView would need to be changed again (removing the filtering code) to allow sqlite to handle the filtering.  Where as if it was decoupled, like sorting, changes would only need to occur in FileData.

I will try to find time soon to do a more thorough review and test it out.

pjft commented:
Hi @jrassa , thanks for taking the time and reading through the code.

That makes perfect sense, and it makes it much cleaner! I'll make that change then and update the branch.

If you do want to test it, functionally it should be fairly solid at the moment, but always appreciate a send pair of eyes. :)

pjft commented:
@jrassa Done.

Thanks for the feedback!

pjft commented:
Hmm. Thinking of it, I might not be supporting filtering AND sorting at the same time now. Will recheck tomorrow.

jrassa commented on es-app/src/guis/GuiGamelistOptions.cpp:
```diff
@@ -60,13 +77,26 @@ GuiGamelistOptions::GuiGamelistOptions(Window* window, SystemData* system) : Gui
+	} 
+	if (mFiltersChanged) 
+	{
+		ViewController::get()->reloadGameListView(mSystem);
```
Calling `reloadGameListView` seems like overkill since it will completely recreate the gamelist view.  Did you try calling  `onFileChanged` like is done above when sorting?

pjft replied:
Oh, sorry, missed this.

I tried this now, and it works for the most part except that it ends up crashing when the list only has a placeholder. I imagine that the getCursor or setCursor methods end up throwing a segmentation fault somewhere.

The performance hit doesn't seem to be tremendous, though.

Given that I am mostly debugging via logs, if you feel this is a dealbreaker, I'm happy to rig this with logs to see where it ends up crashing.

Or, if you have an easy way to troubleshoot it and debug, if you want to let me know where the problem is I'm happy to take a look if the solution isn't obvious to you!

Sorry and thanks for the help here.

pjft replied:
Actually, I tried a bit more and reduced calling the reloadGameListView to only when we're coming from a placeholder to a full list, reason being that otherwise the first selected element did not show any image, video, metadata, etc. Only after I moved.

Every other filtering event turns into a onFileChanged - thanks for the pointer!

Please test to make sure nothing is fully broken - just tried it briefly this morning, but will only be able to test either later or in the coming days.

Thanks.

jrassa commented on es-app/src/FileData.cpp:
```diff
@@ -81,6 +81,37 @@ const std::string& FileData::getThumbnailPath() const
+			// add a placeholder
+			// for now, just create a new FileData of type Folder, with no children
+			// maybe in the future makes sense to create a third type of FileData?
+			FileData* placeholder = new FileData(FOLDER, "<No Results Found for Current Filter Criteria>", mSystem);
```
I would recommend going ahead and creating a PLACEHOLDER type.  You could then removing the placeholder variable and methods from FileData and check against the type instead.

zigurana commented:
Hi guys! Sorry for joining this conversation late, I hope I'm not revisiting passed stations. 
What is the use for the mPlaceholder member? Is it essential to the current design, maybe we can come up with one that does not need it? 
Also, how are empty systems/gamelistviews handled, are they removed completely from the set, or is that what the placeholder is for? 
Finally, will the filter mode be a stored setting? 
Again, this might be treaded ground, if so, apologies.

pjft commented:
Thanks all - and thanks @jrassa for the suggestions!

A few comments, answers and questions:

@zigurana:
1) Placeholder is indeed for the empty gamelists. It is there to display a single line saying that the result list is empty. They are not removed from the set, though - these are not permanent. The filters are only valid within a game system, so you need to be able to go back to the game system and filter by something else. 
2) "Filter mode" is not really a mode per se. It's not a stored setting across sessions - i.e. if you restart ES it'll be unset. That being said, it stays there for the duration of the session - i.e. if you launch a game and return, or if you change system and return to the current one - it stays there. I'm not fully sure whether I'd want to store the filter selection as a whole as a stored setting, I'm thinking about it more as the search function in your email. That being said, nothing would prevent there being a stored session for specific (ideally binary) selections, such as favorites or kids friendly. But right now, storing all the selected options (as it's multi-select, and text based) seems to be an overkill.

To @jrassa 's comments:

Good points, I'll look into that.

In regards to the placeholder type, though, I am still undecided and not thrilled, so I'd love your perspective - in terms of return on effort as well.

Ideally, I'd see the placeholder as not being needed, and the View being able to handle an empty file list. As it stands, right now, it's the FileData list generator that's responsible for NOT providing an empty list, by filling it with a placeholder. Thoughts?

Also, a "Placeholder" FileData type may "subvert" the class purpose.

Should I try to, at render time, handle an empty list properly and fill out the IList with the placeholder item, and make the rest of the code not depend on the IList having FileData type elements in it? Happy to try it, if it isn't too much trouble.

Or alternatively just create a Placeholder FileData type :)

Happy Easter.

zigurana commented:
Hi thanks for those clarifications, that helps a lot.
Of course I am approaching this with the future extensions with Kid/kiosk modes in mind, and in that particular use-case trans-session permanence is a must. Not a big issue, just that a future implementation of the UI modes will likely be something that acts on the current selection of filters, rather than being a filter-type itself.
I would also prefer a situation where empty lists are take care of without the need for a placeholder item. 
The difficulty is mostly in the messy way cursors are currently being used to derive system, root paths, lists of children items,  etc. So a fix of this issue will likely entail a clear (re-)design of what the Systemlists, SystemData, FileData, GameLists, cursor, and the respective views should hold and what not. I've been messing with this a lot the past 18 months, and I still get lost sometimes. 
Maybe we should just take this in small steps, first introduce this filtering with the placeholder, and then, depending on the need and feedback, consider a more fundamental fix. (Agile dev, in a way).

pjft commented:
Hm. Well, trans-session permanence is not hard to implement per se - it's more a matter of deciding the best way to store - and reload - the filter selection. A simple text-based approach leaves too much to be desired, for multi-selections. And we'd need to manage some edge cases, such as what if the metadata changes offline so that the filter selection is no longer valid, etc.

That being said, not rocket science, but just outside of the scope of what I did. Maybe a simple SaveFilterToXML and LoadFilterFromXML type of thingie would work. I'm happy to explore that in the future, should you think it's the right next step for a filtering engine, though I was really looking at it from a "keep it simple yet usable" perspective, especially given the impending (YMMV) metadata re-architecture. Especially as, right now, the filters are a System property, and not a Session property (meaning, you filter individual systems, not a shared filter across systems). Which, truth be told, isn't hard to implement - it's just iterating the collection of filter indexes and setting them all to the same values. I chose not to do so because, as I mentioned in the past, there isn't that much overlap across systems in some categories (namely publishers/developers), and the user experience of seeing half their systems with an empty list didn't seem the best. But open to expanding those - it's easy to just add a checkbox option to "filter all systems" given a specific selection.

I'm all up for small steps, as rewriting the whole rendering and data pipeline is something I am not particularly looking forward to in the short term - too many risks of things that are easily breakable.

That being said, a few simple decisions can be taken now, which are:
- Whether to create a legitimate "Placeholder" type that will later be removed, vs keeping it as a separate boolean flag as it currently is;
- Where to add the placeholder item: at the FileData level, so it never returns an empty list? At rendering, so that it can handle an empty list (even if, at this time, it will just add a placeholder item to it until we have a better architectural decision in place)? Both (i.e. FileData would never return an empty list, but the rendering views would also be better equipped to handle an empty list should one be present)?

Let me know your thoughts.

zigurana commented:
Hey, I agree with all your points. 
I would argue that list-rendering should be made a bit more robust. Having the placeholder be a sort of 'gracefull fallback' when nothing is present would be a good way to solve a lot of similar issues with a single change. 
I would need to do some research on the ramifications of such an approach though, no idea. The iList is a template function right? 

Again, small steps, a FileData placeholder (rather than a more generic type), is a lot smaller in scope. Maybe start with that and see if there is any need for more types / generalization later?

pjft commented:
Happy to create the FileList placeholder for the time being, sure!

Check @jrassa 's recommendations then: would you be fine if I implement these two changes and then we move forward?

jrassa commented:
> Ideally, I'd see the placeholder as not being needed, and the View being able to handle an empty file list. As it stands, right now, it's the FileData list generator that's responsible for NOT providing an empty list, by filling it with a placeholder. Thoughts?

@pjft I think this might be the way to go.  Looking at the code, it looks like TextListComponent should handle an empty list.  `BasicGameListView::populateList` probably needs a slight adjustment to guard against an empty list.  TextListComponent could be updated to have an emptyMessage property that is rendered when the list is empty.

pjft commented:
@jrassa @zigurana 

I did a few things, then:
- Added a PLACEHOLDER type in FileData. I checked the code, and the only place where we seem to make blanket comparisons of (type == GAME ? XXX : YYY) don't matter to this. Where they mattered (if at all) I edited it.
- Removed the responsibility of adding the placeholder in the list to the View, rather than FileData.
- Added a guard in BasicGameListView::populateList against empty lists, and to add a placeholder item to the TextListComponent.

Hope this suits, but would love your thorough testing and thoughts.

zigurana commented:
Good work! I will try to give this a spin somewhere this weekend.

pjft commented:
Thanks.

This isn't actually ready to be merged, as it now has some conflicts with the Go To Random Game PR.

I'm working on making them work together, and rebasing the branch so if you wait a bit I'll let you know when you can fully test it.

Basically, I'm creating a "getDisplayedGameCount()" function and a "getDisplayedFilesRecursive()" one so that you can still go to random games while having some filters on, without risking selecting a hidden file.

Will update when ready.

pjft commented:
Ok, I rebased it and then changed the code to hopefully work well with the Random Game selection.

The bad news is that I still haven't compiled it as I don't have access to my Pi at the moment - only later or in the coming days. 

As such, treat this as untested, and potentially (though hopefully not) not compiling.

Once again, I'll squash the commits when we're all happy with the outcome.

pjft commented:
Tested it, overall looks good. Tested the go to random game with a filtered list, and behaved as expected now.

Let me know your thoughts so hopefully we can close this! 

Have a great weekend.

joolswills commented:
Thanks.  I've not had a chance to test it but will. Any feedback from anyone else is appreciated. Cheers.

joolswills commented:
Thanks.

pjft commented:
Thanks!

@zigurana - let me know if/how you'd want to proceed for the favorites and kids filter implementation. Happy to help.

zigurana commented:
@pjft : hey sorry to leave you hanging like this. I'm away from keyboard for a bit, to get a bit of digital / outdoors balance back in my life. I will have a goid look at these changes, and see how to combine this with the kidmodes abd favorites stuff. Thanks!

pjft commented:
@zigurana Not at all. There's no urgency whatsoever, just wanted to make sure you had a heads up - and I hear that that "outdoors" thing is quite popular. I should look into that as well sometime ;)
Best.

joolswills commented:
I didn't notice this before, but there are indentation issues in es-app/src/FileFilterIndex.cpp from this PR - not sure if other files. some space indentations and trailing whitespace.

pjft commented:
Huh, missed that, sorry. I can push changes for that - my editor is not the
best at highlighting your.

Thanks for noticing.
On Thu, 18 May 2017 at 08:41, Jools Wills <notifications@github.com> wrote:

> I didn't notice this before, but there are indentation issues in
> es-app/src/FileFilterIndex.cpp from this PR - not sure if other files. some
> space indentations and trailing whitespace.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/115#issuecomment-302326536>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAV7b_oTRsY6uhOQN0_SfsbeceXJeJgks5r6_YQgaJpZM4M0OBG>
> .
>

pjft commented:
Oh my - the horror.  I pretty much removed trailing spaces from 19 of the 20 files. But now at least I have my editor set up to handle that.

About the indentation issues, though, I can't seem to find them after two passes, though I may be missing something obvious on my end. Maybe it's a matter of spaces vs tabs and how my editor renders them.

Could you point me in the right direction? And would you rather have this as a separate PR?
